hail/python/hailtop/hailctl/batch/submit.py

```diff
@@ -46,14 +64,17 @@ def file_input_to_src_dest(file: str) -> Tuple[str, str, str]:
         src = result.get('src')
         if src is None:
             raise ValueError(f'invalid file specification {file}. Must have a "src" defined.')
-        src = os.path.abspath(os.path.expanduser(src))
+        src = real_absolute_expanded_path(src)
         src = src.rstrip('/')
 
         dest = result.get('dest')
         if dest is not None:
-            dest = os.path.abspath(os.path.expanduser(dest))
+            dest_intended_as_directory = dest[-1] == '/'
+            dest = real_absolute_expanded_path(dest)
+            if dest_intended_as_directory:
+                dest = os.path.join(dest, os.path.basename(src))
         else:
-            dest = os.getcwd()
+            dest = os.path.join(real_absolute_cwd(), os.path.basename(src))
```

Comment:
@jigold commented at https://github.com/hail-is/hail/pull/14183:

> Can you explain this change?

The semantics of `ln -s src dst` depends on whether `dst` is a directory or not. I found it easier to reason about `file_input_to_src_dest` when I had the post-condition: for each triplet, the destination is the desired location of a symlink to the source.

Before this change the post-condition was "the destination is the desired location of a symlink to the source or a folder into which we want a symlink placed". The latter situation feels like a footgun. The presence or absence of a trailing slash (which realname removes) affects the semantics. Moreover `basename /foo/` prints `foo` whereas `os.path.basename('/foo/')` returns `''`.

With these changes, we still use the trailing slash to determine if we want copy-to or copy-into, but once that's resolved, we're always dealing with fully specified source-destination file-pairs.